### PR TITLE
[BEAM-3099] Add and migrate to HdfsCLI library for Python SDK.

### PR DESCRIPTION
**Pull request by udim**
With this PR, HDFS should be ready for input.
For output, we still need to resolve BEAM-3600 first.

### CODE REVIEW

**Comment by udim** [comment]
R: @chamikaramj

File: sdks/python/apache_beam/options/pipeline_options.py [hunk 1]
```diff
@@ -392,6 +393,33 @@ def validate(self, validator):
     return errors
 
 
+class HdfsOptions(PipelineOptions):
```
Comment (chamikaramj) [inline sdks/python/apache_beam/options/pipeline_options.py]: Let's call this HadoopFileSystemOptions to match the name of the file-system.

File: sdks/python/apache_beam/io/filesystemio.py [hunk 2]
```diff
@@ -158,7 +158,7 @@ def readable(self):
 
 
 class UploaderStream(io.RawIOBase):
-  """Provides a stream interface for Downloader objects."""
+  """Provides a stream interface for Uploader objects."""
```
[the remark this answers is not in the document]
Comment (udim) [reply sdks/python/apache_beam/io/filesystemio.py]: I thought the name was apt considering that it generalizes a part of gcsio.
If you feel strongly about this I could rename it in another PR.

File: sdks/python/apache_beam/io/hadoopfilesystem.py [hunk 3]
```diff
@@ -20,45 +20,93 @@
...
+  def __init__(self, hdfs_client, path):
+    self._hdfs_client = hdfs_client
+    self._path = path
+    self._size = self._hdfs_client.status(path)['size']
+
+  @property
+  def size(self):
+    return self._size
+
+  def get_range(self, start, end):
```
[the remark this answers is not in the document]
Comment (udim) [reply sdks/python/apache_beam/io/hadoopfilesystem.py]: This method will be used by DownloaderStream. I don't understand the rest of your comment - does it have to do with buffering? If HdfsDownloader or the underlying client read past end that's an implementation detail.

File: sdks/python/apache_beam/io/hadoopfilesystem.py [hunk 4]
```diff
@@ -20,45 +20,93 @@
...
 
 
-# TODO(udim): Add @retry.with_exponential_backoff to some functions, like in
-# gcsio.py.
+class HdfsDownloader(filesystemio.Downloader):
+
+  def __init__(self, hdfs_client, path):
+    self._hdfs_client = hdfs_client
+    self._path = path
+    self._size = self._hdfs_client.status(path)['size']
```
[the remark this answers is not in the document]
Comment (udim) [reply sdks/python/apache_beam/io/hadoopfilesystem.py]: I'd rather leave this where it is. This invocation also serves as verification that the file exists.

File: sdks/python/apache_beam/io/hadoopfilesystem.py [hunk 5]
```diff
@@ -123,11 +171,14 @@ def match(self, url_patterns, limits=None):
           'Patterns and limits should be equal in length: %d != %d' % (
               len(url_patterns), len(limits)))
 
-    # TODO(udim): Update client to allow batched results.
     def _match(path_pattern, limit):
       """Find all matching paths to the pattern provided."""
-      file_infos = self._hdfs_client.ls(path_pattern, detail=True)[:limit]
-      metadata_list = [FileMetadata(file_info['name'], file_info['size'])
+      fs = self._hdfs_client.status(path_pattern, strict=False)
+      if fs and fs['type'] == 'FILE':
```
[the remark this answers is not in the document]
Comment (udim) [reply sdks/python/apache_beam/io/hadoopfilesystem.py]: Done.

File: sdks/python/apache_beam/io/hadoopfilesystem.py [hunk 6]
```diff
@@ -123,11 +171,14 @@ def match(self, url_patterns, limits=None):
...
+        file_infos = [(fs['pathSuffix'], fs)][:limit]
```
[the remark this answers is not in the document]
Comment (udim) [reply sdks/python/apache_beam/io/hadoopfilesystem.py]: Done, as module level variables.

File: sdks/python/apache_beam/options/pipeline_options.py (hunk 1 above)
Comment (udim) [reply sdks/python/apache_beam/options/pipeline_options.py]: Done.

**Comment by udim** [comment]
PTAL

File: sdks/python/apache_beam/io/hadoopfilesystem.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (chamikaramj) [reply sdks/python/apache_beam/io/hadoopfilesystem.py]: Got it. Current usage is fine. I was pointing out that reading exact byte range [a, b) is usually not what is needed as long as sources are concerned.

File: sdks/python/apache_beam/io/filesystemio.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chamikaramj) [reply sdks/python/apache_beam/io/filesystemio.py]: Given that this is a set of utils not expected to be directly used by end users I think it makes sense to rename this. A separate PR is fine.

**Comment by chamikaramj** [comment]
Run Python PreCommit

**Comment by udim** [comment]
Regard test failures, 
apache_beam.testing.test_stream_test:TestStreamTest.test_gbk_execution
is currently broken at head.

**Comment by udim** [comment]
Run Python PreCommit

**Pull request merged**